metaworld/envs/mujoco/sawyer_xyz/sawyer_xyz_env.py

```diff
@@ -426,3 +435,78 @@ def _get_state_rand_vec(self):
             self._last_rand_vec = rand_vec
             return rand_vec
+
+    def _gripper_caging_reward(self, action, obj_position, obj_radius):
+        pad_success_margin = 0.05 # obj_radius + 0.01
+        grip_success_margin = obj_radius + 0.01
+        x_z_success_margin = 0.005
+
+        tcp = self.tcp_center
+        left_pad = self.get_body_com('leftpad')
+        right_pad = self.get_body_com('rightpad')
+        delta_object_y_left_pad = left_pad[1] - obj_position[1]
+        delta_object_y_right_pad = obj_position[1] - right_pad[1]
+        right_caging_margin = abs(abs(obj_position[1] - self.init_right_pad[1]) - pad_success_margin)
+        left_caging_margin = abs(abs(obj_position[1] - self.init_left_pad[1]) - pad_success_margin)
+
+        right_caging = reward_utils.tolerance(delta_object_y_right_pad,
+            bounds=(obj_radius, pad_success_margin),
+            margin=right_caging_margin,
+            sigmoid='long_tail',
+        )
+        left_caging = reward_utils.tolerance(delta_object_y_left_pad,
+            bounds=(obj_radius, pad_success_margin),
+            margin=left_caging_margin,
+            sigmoid='long_tail',
+        )
+
+        right_gripping = reward_utils.tolerance(delta_object_y_right_pad,
+            bounds=(obj_radius, grip_success_margin),
+            margin=right_caging_margin,
+            sigmoid='long_tail',
+        )
+        left_gripping = reward_utils.tolerance(delta_object_y_left_pad,
+            bounds=(obj_radius, grip_success_margin),
+            margin=left_caging_margin,
+            sigmoid='long_tail',
+        )
+
+
+        assert right_caging >= 0 and right_caging <= 1
+        assert left_caging >= 0 and left_caging <= 1
+
+        y_caging = reward_utils.hamacher_product(right_caging, left_caging)
+        y_gripping = reward_utils.hamacher_product(right_gripping, left_gripping)
+
+        assert y_caging >= 0 and y_caging <= 1
+
+        tcp_xz = tcp + np.array([0., -tcp[1], 0.])
+        obj_position_x_z = np.copy(obj_position) + np.array([0., -obj_position[1], 0.])
+        tcp_obj_norm_x_z = np.linalg.norm(tcp_xz - obj_position_x_z, ord=2)
+        init_obj_x_z = self.obj_init_pos + np.array([0., -self.obj_init_pos[1], 0.])
+        init_tcp_x_z = self.init_tcp + np.array([0., -self.init_tcp[1], 0.])
+
+
+        tcp_obj_x_z_margin = np.linalg.norm(init_obj_x_z - init_tcp_x_z, ord=2) - x_z_success_margin
+        x_z_caging = reward_utils.tolerance(tcp_obj_norm_x_z,
+                                bounds=(0, x_z_success_margin),
+                                margin=tcp_obj_x_z_margin,
+                                sigmoid='long_tail',)
+
+        assert right_caging >= 0 and right_caging <= 1
+        gripper_closed = min(max(0, action[-1]), 1)
+        assert gripper_closed >= 0 and gripper_closed <= 1
+        caging = reward_utils.hamacher_product(y_caging, x_z_caging)
+        assert caging >= 0 and caging <= 1
+
+        if caging > 0.95:
```

Comment:
how well does this work out for you? I'm trying it out in one of my drawer prs just to see, and  I get good caging but grasping still fails.

Reply:
this works very consistently for me. the one thing is your object radius needs to be correct, or even slightly less than the true object radius. if not it just cages 'closer' to the object than the old one but not yet exactly gripping it